Slurp and serve UTF-8 encoded files

landakram opened the pull request:
Hi @magnars! First off, thank you for this incredibly useful library.

When I tried to serve files pulled using `slurp-directory`, I noticed that they did not correctly render UTF-8 encoded characters (instead rendering a `?`). 

This PR adds the ability to pass `opts` into `slurp-directory`, which are then passed to underlying calls to `slurp`. This can be used to pass an encoding like `(slurp-directory dir pattern :encoding "UTF-8")`.

It also adds the UTF-8 charset to the `Content-type` header so that files served through ring can be rendered correctly, and fixes tests to reflect this change.

Let me know if there's anything I'm missing and have a great day.

magnars commented:
Thanks for the pull request. :) I'm a little baffled at why my own utf-8 chars are working fine, but I guess it's using some system defaults when nothing is specified. It's the classic "works on my machine"!

Since we're now mandating utf-8, is this a breaking change? Should `opts` to `slurp-directory` be `:encoding "UTF-8"` by default?

landakram commented:
Huh — playing with it more, it looks like when I don't include `:encoding "UTF-8"`, the file contents actually print correctly in my shell, but make it to the browser with the `?` characters. So the `opts` change is actually unnecessary. I can remove it if you like, but it's also nice to be able to pass `opts` through to `slurp`. 

I don't think we should add `:encoding "UTF-8"` to `opts` in `slurp-directory` by default, unless it's conditional on the `:encoding` key not already being there. That's because I imagine people are more likely to use `slurp-directory` with their own `:encoding`, independently of `serve-pages` (for instance, if they skip serving and go straight to building the static site). It's probably best to let them specify what works with their files.

magnars commented:
Would `{"Content-Type" "text/html; charset=utf-8"}` still be correct if people are using different `:encoding` to slurp tho?

landakram commented:
To be honest, I'm not totally sure what would be displayed.

I just realized, however, that setting the charset can be done with the
[content-type or charset middleware](
https://github.com/mmcgrana/ring/blob/master/ring-core/src/ring/util/response.clj#L130),
which means that forcing the default of utf-8 isn't necessary in stasis
itself (I can just do it in my own code).

Should I remove the charset logic from serve-pages? I still think that
changing slurp-directory to pass opts though is useful.

On Mon, Dec 21, 2015 at 2:16 AM Magnar Sveen <notifications@github.com>
wrote:

> Would {"Content-Type" "text/html; charset=utf-8"} still be correct if
> people are using different :encoding to slurp tho?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/magnars/stasis/pull/15#issuecomment-166260518>.
>

magnars commented:
I was going to suggest adding a note about the content-type middleware to the readme, but I see that it already includes `wrap-content-type` in one example. Maybe it needs to be more explicit in the Q & A section. :)

So yeah, let's not mandate a content type on the header. Passing options to slurp sounds like a good thing regardless.

landakram commented:
Yes I didn't see that :) 

Even after I added `wrap-content-type` though, I also had to include `ring-headers` and use [`wrap-default-charset`](https://ring-clojure.github.io/ring-headers/ring.middleware.default-charset.html), since on my system it was using `ISO-8859-1` by default. Super weird.

I've extracted the opts logic into https://github.com/magnars/stasis/pull/16, so I'm closing this.
